File: sample/send_to_windows/Sources/TransferCore/Naming.swift
// Naming.swift — §4.3 네이밍 규칙.
//
//   <YYYY-MM-DD>__<category-key>__<basename>__v<NN><ext>
//
// 규칙:
// - category-key는 영문 소문자 (한글 라벨 X)
// - basename은 원본 파일/폴더 이름의 확장자 제외 부분 그대로 (한글/공백 허용)
// - 폴더 전송 시 ext 없음 (빈 문자열)
// - basename은 NFC 정규화 (HFS+/SMB NFD 변환 이슈 회피)
// - 버전 v01~v99 (2자리 zero pad)

import Foundation

public enum Naming {

    public struct Components: Sendable, Equatable {
        public let date: String        // "2026-05-18"
        public let categoryKey: String // "documents"
        public let basename: String    // "예상 공수 산정 근거" (NFC)
        public let version: Int        // 1
        public let ext: String         // ".html" or "" for folder

        public init(date: String, categoryKey: String, basename: String, version: Int, ext: String) {
            self.date = date
            self.categoryKey = categoryKey
            self.basename = basename
            self.version = version
            self.ext = ext
        }
    }

    /// 사용자 원본 경로(URL)에서 basename + ext 추출.
    /// 폴더면 ext = "", basename = 폴더명.
    /// 파일이면 ext = lastPathComponent의 마지막 "."부터 끝까지 (점 포함).
    public static func split(originalURL url: URL,
                             isDirectory: Bool) -> (basename: String, ext: String) {
        let name = url.lastPathComponent.precomposedStringWithCanonicalMapping
        if isDirectory {
            return (name, "")
        }
        // 마지막 '.' 위치. 첫 글자가 '.'인 dotfile은 ext 없는 것으로 취급.
        if let dotIdx = name.lastIndex(of: "."), dotIdx != name.startIndex {
            let basename = String(name[..<dotIdx])
            let ext = String(name[dotIdx...])
            return (basename, ext)
        }
        return (name, "")
    }

    /// 컴포넌트 → 파일명 문자열.
    public static func render(_ c: Components) -> String {
        let v = String(format: "v%02d", c.version)
        return "\(c.date)__\(c.categoryKey)__\(c.basename)__\(v)\(c.ext)"
    }

    /// 편의: date + 카테고리 + 원본 URL + 버전 → 최종 파일명.
    public static func render(date: Date = .init(),
                              category: TransferCategory,
                              originalURL: URL,
                              isDirectory: Bool,
                              version: Int = 1,
                              timeZone: TimeZone = .current) -> String {
        let (basename, ext) = split(originalURL: originalURL, isDirectory: isDirectory)
        let c = Components(
            date: Timestamps.filenameDate(date, timeZone: timeZone),
            categoryKey: category.key,
            basename: basename,
            version: version,
            ext: ext
        )
        return render(c)
    }

    /// 파일명 역파싱. 형식 안 맞으면 nil.
    /// 주의: basename에 "__" 포함될 수 있어서 (예: "my__report"), 끝에서부터 토큰화.
    public static func parse(_ filename: String) -> Components? {
        let nfc = filename.precomposedStringWithCanonicalMapping

        // ext 분리 (basename 토큰화 전에)
        var stem: String
        var ext: String
        if let dotIdx = nfc.lastIndex(of: "."), dotIdx != nfc.startIndex {
            stem = String(nfc[..<dotIdx])
            ext  = String(nfc[dotIdx...])
        } else {
            stem = nfc
            ext  = ""
        }

        // 끝에서 "__v<NN>" 추출
        guard let vRange = stem.range(of: #"__v\d{2}$"#, options: .regularExpression) else {
            return nil
        }
        let vToken = String(stem[vRange]).dropFirst(3)   // "v01" → "01"
        guard let version = Int(vToken) else { return nil }
        stem = String(stem[..<vRange.lowerBound])

        // 앞에서 "<date>__<category>__" 추출
        let parts = stem.components(separatedBy: "__")
        guard parts.count >= 3 else { return nil }
        let date = parts[0]
        let categoryKey = parts[1]
        let basename = parts[2...].joined(separator: "__")

        // 형식 검증
        guard date.range(of: #"^\d{4}-\d{2}-\d{2}$"#, options: .regularExpression) != nil else {
            return nil
        }
        guard !categoryKey.isEmpty, !basename.isEmpty else { return nil }

        return Components(date: date, categoryKey: categoryKey,
                          basename: basename, version: version, ext: ext)
    }
}

File: mac_gui/share-manager/src-tauri/src/transfer/naming.rs
// naming.rs — port of Naming.swift. §4.3 filename rule:
//
//   <YYYY-MM-DD>__<category-key>__<basename>__v<NN><ext>
//
// - category_key is the english lower-case key (not the Korean label)
// - basename is the original stem (any Unicode), NFC-normalized
// - directories have no ext (empty string)
// - version is v01..v99 (zero-padded)

use chrono::{DateTime, Local};
use std::path::Path;
use unicode_normalization::UnicodeNormalization;

use super::timestamps;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Components {
    pub date: String,
    pub category_key: String,
    pub basename: String,
    pub version: u32,
    pub ext: String,
}

/// (basename, ext) split. ext = "" for directories and dotfiles without a
/// secondary extension; otherwise ".ext" (leading dot included).
pub fn split(name: &str, is_directory: bool) -> (String, String) {
    let nfc: String = name.nfc().collect();
    if is_directory {
        return (nfc, String::new());
    }
    // Skip leading-dot files: dotfiles without a secondary dot keep ext = "".
    if let Some(idx) = nfc.rfind('.') {
        if idx > 0 {
            let (b, e) = nfc.split_at(idx);
            return (b.to_string(), e.to_string());
        }
    }
    (nfc, String::new())
}

pub fn render_components(c: &Components) -> String {
    format!(
        "{date}__{key}__{base}__v{ver:02}{ext}",
        date = c.date,
        key  = c.category_key,
        base = c.basename,
        ver  = c.version,
        ext  = c.ext,
    )
}

pub fn render(
    date: DateTime<Local>,
    category_key: &str,
    original: &Path,
    is_directory: bool,
    version: u32,
) -> String {
    let name = original
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or("");
    let (basename, ext) = split(name, is_directory);
    render_components(&Components {
        date: timestamps::filename_date(date),
        category_key: category_key.to_string(),
        basename,
        version,
        ext,
    })
}

/// Reverse-parse a rendered filename. Returns None if shape doesn't match.
/// Basenames may contain "__", so we tokenize from the tail.
pub fn parse(filename: &str) -> Option<Components> {
    let nfc: String = filename.nfc().collect();

    let (stem, ext) = if let Some(idx) = nfc.rfind('.') {
        if idx > 0 {
            (nfc[..idx].to_string(), nfc[idx..].to_string())
        } else {
            (nfc.clone(), String::new())
        }
    } else {
        (nfc.clone(), String::new())
    };

    // tail: "__v<NN>"
    let tail_len = "__v00".len();
    if stem.len() < tail_len { return None; }
    let (head, tail) = stem.split_at(stem.len() - tail_len);
    if !tail.starts_with("__v") { return None; }
    let ver_str = &tail[3..];
    if ver_str.len() != 2 || !ver_str.chars().all(|c| c.is_ascii_digit()) {
        return None;
    }
    let version: u32 = ver_str.parse().ok()?;

    let parts: Vec<&str> = head.splitn(3, "__").collect();
    if parts.len() < 3 { return None; }
    let (date, category_key, basename) = (parts[0], parts[1], parts[2]);

    // shape check: YYYY-MM-DD
    if date.len() != 10
        || !date.chars().enumerate().all(|(i, c)| match i {
            4 | 7 => c == '-',
            _ => c.is_ascii_digit(),
        })
    {
        return None;
    }
    if category_key.is_empty() || basename.is_empty() { return None; }

    Some(Components {
        date: date.to_string(),
        category_key: category_key.to_string(),
        basename: basename.to_string(),
        version,
        ext,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_canonical_filename() {
        let c = Components {
            date: "2026-05-18".to_string(),
            category_key: "documents".to_string(),
            basename: "예상 공수 산정 근거".to_string(),
            version: 1,
            ext: ".html".to_string(),
        };
        assert_eq!(render_components(&c), "2026-05-18__documents__예상 공수 산정 근거__v01.html");
    }

    #[test]
    fn parse_round_trip() {
        let s = "2026-05-18__documents__my__report__v07.html";
        let c = parse(s).expect("parse");
        assert_eq!(c.date, "2026-05-18");
        assert_eq!(c.category_key, "documents");
        assert_eq!(c.basename, "my__report");
        assert_eq!(c.version, 7);
        assert_eq!(c.ext, ".html");
    }

    #[test]
    fn folder_has_no_ext() {
        let (b, e) = split("my-folder", true);
        assert_eq!(b, "my-folder");
        assert_eq!(e, "");
    }

    #[test]
    fn dotfile_has_no_ext() {
        let (b, e) = split(".gitignore", false);
        assert_eq!(b, ".gitignore");
        assert_eq!(e, "");
    }
}
